### ISSUE TITLE
Embedded jetty doesn't set the charset to content-type header

### ISSUE THREAD

**Issue by wodeidaiga** [eclipse/jetty.project#11289]
**Jetty Version**
12.0.3
**Jetty Environment**

ee10
**Java Version**
17
**Question**
When using Embedded Jetty, the `setContentType()` method in the class `ServletApiResponse` won't set the charset=utf8 automatically like the jetty 9 does. Is there any configuration needed to let jetty set it?
Thanks!

**Comment by joakime** [comment]
We have plenty of unit tests that test this exact scenario.
You might have discovered a corner case we are missing.

Please detail exactly how you are using `HttpServletResponse.setContentType()`, what parameters you are using, and what other methods on `HttpServletResponse` using (in the order you use them)

**Comment by wodeidaiga** [comment]
@joakime ,
The cause is because of the following code:
![image](https://github.com/jetty/jetty.project/assets/9336402/0655e352-6ba5-4156-a151-b8277a5a9816)
![image](https://github.com/jetty/jetty.project/assets/9336402/c880917e-d0b4-4a58-93c7-730da995cf8a)
If we call the getoutputstream then run setContentType(), the charset will not be set because the iswriting() returns false

**Comment by joakime** [comment]
Please give an example of _how **you** are using the HttpServletResponse_.

You are just pointing out a normal servlet spec behavior in our code.

The calls to `.getOutputStream()` and `.getWriter()` should only be called AFTER you set all of the response headers.

That's because those calls use the response headers to know how to setup those streams.
The minute you call either `.getOutputStream()` or `.getWriter()` you are forbidden (by the servlet spec) from changing headers required for those 2 method calls.

Also, the `isWriting()` test is related to detecting if the servlet output type is currently in WRITER (vs STREAM), if you called `.getOutputStream()` then `isWriting()` is definitely false (and is following the servlet spec).

In other words, each of these examples is a failure due to bad usage of the Servlet API.

``` java
// Example 1
PrintWriter out = response.getWriter();
response.setContentType("text/plain"); // this should be done before .getWriter().

// Example 2
PrintWriter out = response.getWriter();
response.setCharacterEncoding("utf-8"); // this should be done before .getWriter();

// Example 3
ServletOutputStream stream = response.getOutputStream();
response.setContentType("application/json; charset=utf-8"); // this should be done before .getOutputStream(), 
// also charset is ignored in this case as application/json is by-spec not allowed to have a charset.

// Example 4
ServletOutputStream stream = response.getOutputStream();
PrintWriter out = response.getWriter(); // this is a failure in the servlet API (output type has been set to STREAM already)

// Example 5
PrintWriter out = response.getWriter();
ServletOutputStream stream = response.getOutputStream();  // this is a failure in the servlet API (output type has been set to WRITER already)
```

**Comment by gregw** [comment]
@joakime you are able to set the charset and character encoding after `getOutputStream`, but only before the response is committed.  It is still a strange/fragile order to do things in.
I have written the follow test to show this working:
```
    @Test
    public void testCharsetAfterGetOutputStream() throws Exception
    {
        ServletContextHandler contextHandler = new ServletContextHandler();
        contextHandler.setContextPath("/");

        HttpServlet charsetResetToJsonMimeTypeServlet = new HttpServlet()
        {
            @Override
            protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException
            {
                OutputStream out = response.getOutputStream();
                response.setCharacterEncoding("utf-8");
                response.setContentType("text/html");
                out.write("hello".getBytes(StandardCharsets.UTF_8));
            }
        };

        contextHandler.addServlet(charsetResetToJsonMimeTypeServlet, "/charset/hello/*");
        startServer(contextHandler);

        HttpTester.Request request = new HttpTester.Request();
        request.setMethod("GET");
        request.setURI("/charset/hello/");
        request.setVersion(HttpVersion.HTTP_1_1);
        request.setHeader("Connection", "close");
        request.setHeader("Host", "test");

        ByteBuffer responseBuffer = connector.getResponse(request.generate());
        // System.err.println(BufferUtil.toUTF8String(responseBuffer));
        HttpTester.Response response = HttpTester.parseResponse(responseBuffer);

        // Now test for properly formatted HTTP Response Headers.
        assertThat("Response Code", response.getStatus(), is(200));
        // The Content-Type should not have a charset= portion
        assertThat("Response Header Content-Type", response.get("Content-Type"), is("text/html;charset=UTF-8"));
    }
```

@chiqiu the `isWriting` check in the code you suggest is correct.  The response is not writing.  If it were, then it would be too late to set the charset, as the writer would already have a charset locked in.  It is line 654 in the file you reference that sets the charset for the output stream case.

**Comment by wodeidaiga** [comment]
@joakime @gregw ,
Do you mean that we must call setHeader or setContentType before call the getOutputStream ?  I checked that the line 654 will not be called if call getOutputSteam first :
![image](https://github.com/jetty/jetty.project/assets/9336402/8b3b9399-4d87-4075-887a-9e46b16c5be5)
Will this be fixed or some status check can be made to add restrictions to the call order?
Thanks

**Comment by gregw** [comment]
@chiqiu you don't have to set content-type before calling getOutputStream, but it is the more normal (less fragile) thing to do, as if you use the output stream after getting it and before setting the content-type, then the response will be committed and you will no longer be able to set the content-type.

At this stage, you have not described anything that is a problem, so there is nothing to be fixed.

Can you explain a bit more what you mean by "set the charset=utf-8 automatically"?

Specifically what calls are you making to either/or: setCharacterEncoding, setContentType; and what arguments are you passing them?

Could you perhaps write a small test servlet that demonstrates what you think the problem is?

**Comment by wodeidaiga** [comment]
@gregw ,
The issue is we have missing charset header on 12.0x.  I wrote a test servlet :
If we run it on Jetty 9.4, the return Content-Type header value includes the 'charset=utf-8'.
If we run it on Jetty 12.0x, the return Content-Type header doesn't have the 'charset=utf8'.
The order is setCharacterEncoding -> getStream -> setContentType
This can cause behavior change on our client software.
Since we are migrating from 9.4 we expect the same return values.

``` java
@WebServlet(urlPatterns = {"/test"})
public class TestServlet extends HttpServlet
{
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException
    {
        response.setCharacterEncoding("utf-8");
        OutputStream outputStream = response.getOutputStream();
        response.setContentType("application/json");
        outputStream.write("{}".getBytes(StandardCharsets.UTF_8));
        outputStream.close();
    }
}
```

Thanks

**Comment by janbartel** [comment]
@chiqiu there may well be a difference between jetty-9 and jetty-12, as we did some work in jetty-12 to ensure compliance with the servlet spec character and content type encoding requirements. Furthermore, we overhauled the mimetypes. `application/json` is defined to be encoded by default in `UTF-8` (with `UTF-16` or `UTF-32` automatically detectable) as specified in https://www.ietf.org/rfc/rfc4627.txt section 3 `Encoding`.  Furthermore, according to IANA no charset should be sent with this mimetype:  https://www.iana.org/assignments/media-types/application/json.

So in short, yes, there is a difference with jetty-9, but jetty-12 is correct.

**Comment by wodeidaiga** [comment]
@janbartel ,
If we change the order of calling the APIs above to  setContentType->setCharacterEncoding the charset header will be set automatically. Seems jetty 12 is also trying to add it if possible as what Jetty 9 does.
We have legacy http client code which handles the charset value, because it was set with the content type before. I think there may also be users have a situation like ours. Could you please see if it can be added as in the Jetty 9 ?
Thanks!

**Comment by janbartel** [comment]
@chiqiu  if you really need the charset to be in the response headers, you can do it if you set the _entire_ content type string like so:

```java
  response.setContentType("application/json;charset=utf-8");
```

**Comment by wodeidaiga** [comment]
> @chiqiu if you really need the charset to be in the response headers, you can do it if you set the _entire_ content type string like so:
> 
> ```java
>   response.setContentType("application/json;charset=utf-8");
> ```
Thanks, I will try your suggestion.
The thing is that we are using a legacy REST framework which sets the content-type, so we have to write a HttpResponseWrapper to set it. The other reason I open this issue is to see if there is potential issues since it is some jetty behavior change.

**Comment by gregw** [comment]
@chiqiu what happens if you run your application in the ee8 or ee9 environment?

Actually if it is legacy code, how are you running a jetty-9 webapp on ee10? as the name space has changed from javax to jakarta.

If you see different behaviour between jetty-10 and ee8 or jetty-11 and ee9, then we may fix this, as we want the older ee environments to be bug for bug compatible.

**Comment by wodeidaiga** [comment]
> older

@gregw ,
I haven't tested with ee8 or ee9 on Jetty 12. I will check if it is different.
We used tools to upgraded the java imports in the legacy framework and recompiled it but we didn't change the other codes. The framework sets "application/json" by default and has some implementation to adjust the charsets which I think is risky to be modified. 

Thanks

**Comment by joakime** [comment]
> @chiqiu there may well be a difference between jetty-9 and jetty-12, as we did some work in jetty-12 to ensure compliance with the servlet spec character and content type encoding requirements. Furthermore, we overhauled the mimetypes. `application/json` is defined to be encoded by default in `UTF-8` (with `UTF-16` or `UTF-32` automatically detectable) as specified in https://www.ietf.org/rfc/rfc4627.txt section 3 `Encoding`. Furthermore, according to IANA no charset should be sent with this mimetype: https://www.iana.org/assignments/media-types/application/json.
> 
> So in short, yes, there is a difference with jetty-9, but jetty-12 is correct.

Slightly out of date on those links there (RFC4627 is out of date).

* IANA Registration Master Table for media types - https://www.iana.org/assignments/media-types/media-types.xhtml
* IANA Registration for `application/json` - https://www.iana.org/assignments/media-types/application/json
* IANA Registration link for up to date spec on `application/json` - https://www.rfc-editor.org/rfc/rfc8259.html

The up to date RFC8259 points out specifically in Section 11 - https://www.rfc-editor.org/rfc/rfc8259.html#section-11

> ```
>    Note:  No "charset" parameter is defined for this registration.
>      Adding one really has no effect on compliant recipients.
> ```

**Comment by joakime** [comment]
> I haven't tested with ee8 or ee9 on Jetty 12. I will check if it is different.
> We used tools to upgraded the java imports in the legacy framework and recompiled it but we didn't change the other codes. The framework sets "application/json" by default and has some implementation to adjust the charsets which I think is risky to be modified.

@chiqiu the `charset` parameter is completely ignored by JSON parsers.
They all parse in UTF-8, all the time.

Here's the history.

* July 2006 - https://datatracker.ietf.org/doc/html/rfc4627#section-3 - Original spec - allowed pretty much anything encoding wise.
* March 2014 - https://datatracker.ietf.org/doc/html/rfc7159#section-8.1 - Updated spec - allowed only UTF-8, UTF-16, and UTF-32 but only via auto-detection (the `charset` attribute on the `Content-Type` was not to be used)
* December 2017 - https://datatracker.ietf.org/doc/html/rfc8259#section-8.1 - Updated spec - no longer allows anything but UTF-8 on the public internet, `charset` has no meaning. (closed systems can do their own thing)

A few parsers used to have the ability to alter the charset, but most of that functionality has been removed over the past 7 years as it is incompatible on the public internet, where JSON *MUST* (according to spec) be UTF-8 and nothing else.    If you are on an isolated network, with no public internet use of any kind, then you have a legit use case for charset outside of UTF-8.  But in that last case you will be stuck with old versions of everything to support that behavior.

Jetty added the assumed utf-8 charset on `application/json` use back in March 2018 to support the proper to-spec behavior with `application/json` (ie: with no charset).

**Comment by wodeidaiga** [comment]
@joakime ,
Thank you for your explanation. We have some issue with the legacy client codes some of them which might be very hard to upgrade. I will try to fix it from our app side. My biggest concern is that I might miss something during the upgrade. Now it is clear to me.
Thanks!

**Comment by flozano** [comment]
I have a webapp that sets `application/json` everywhere, and I did add `setCharacterEncoding("utf-8")`, and previously I was obtaining `application/json;charset=utf-8` in my responses.

This explanation about json being always utf-8 makes total sense from a standards perspective and "correct" client behaviors... but I would like to keep backward compatibility: I have devices out there with a fixed firmware that I don't control, and therefore I would like to minimize the changes in my responses when doing any server change. What would be the best way to handle this in jetty 12 / EE10, other than actually modifying my application to include the explicit charset when setting content-type?

**Comment by gregw** [comment]
I'm reopening this, as there should be a way to force jetty to emit the content-type with charset, even if just to support legacy clients.

**Comment by gregw** [comment]
@flozano Have you tried just adding an explicit `response.setHeader("Content-Type", "application/json;charset=utf8")`?

**Comment by flozano** [comment]
yes, that's the plan B. but it's a big app and modifying a ton of MVC controllers is not ideal.

I was wondering if any place in the container can allow overriding it in a centralized manner.

**Comment by gregw** [comment]
You can wrap the response in a filter and do that on setContentType.   But you'd really want to detect old/legacy clients and only do that conditionally.  If you do it always, then you do it forever!

**Comment by flozano** [comment]
that's what I'm going to do for now, wrap the response in a wrapper that handles the case in setHeader and in setContentType. but it would be great if such thing was possible on jetty directly, just as it worked before and just as it works on other containers (from which I'm migrating my code). not by default.


about the clients... 

I don't have any control over the devices. One can safely assume browsers get up to date, but when using http to expose APIs for non-browser devices, funny things happen when people implement client-side and implicit contracts are created/assumed, and that's something some projects must navigate over the years. unfortunately.

thanks for the response 👍🏻

**Comment by gregw** [comment]
@flozano I'm leaning towards being convinced... if the app calls `setContentType("application/json;charset=utf8")`, then that's the header we should send.  Even if it is not correct and ignored by all modern clients.

@joakime what is the argument for us rewriting this header?
We don't correct "text/plane" to "text/plain", so why should we correct other mistakes? 

I think the removal of the charset is probably a side effect of our optimization to find a pre-encoded field, but optimizations should not change behaviour (or at least at least not if possible).

I'll have a look at this soon....

**Comment by gregw** [comment]
Oh but I'm not convinced to add a charset to the content type if it is not passed explicitly.  So in the following case:
```java
        response.setCharacterEncoding("utf-8");
        response.setContentType("application/json");
```
we would not add the charset, because json is utf-8.

**Comment by flozano** [comment]
OK... I *believe* this `setContentType("application/json;charset=utf8"` works already ootb.... shouldn't it? I was assuming so. I will create unit test in my embedded setup to confirm.

**Comment by flozano** [comment]
> we would not add the charset, because json is utf-8.

I see. Thanks for considering it anyway 👍

**Comment by flozano** [comment]
> OK... I believe this setContentType("application/json;charset=utf8" works already ootb.... shouldn't it? I was assuming so. I will create unit test in my embedded setup to confirm.

I confirmed: we have tests for a default error servlet which sets:

```java
		res.setCharacterEncoding("UTF-8");
		res.setHeader("Cache-Control", CacheControl.DEFAULT);
		res.setContentType("application/json;charset=utf-8");
```
and this test passes with jetty:

```java
		try (CloseableHttpClient client = HttpClients.createDefault()) {
			try (CloseableHttpResponse response = client.execute(new HttpGet(server.getBaseURL() + "non-mapped-url"))) {
				assertNull(response.getHeader("Vary"));
				assertEquals(404, response.getCode());
				assertEquals( "application/json;charset=utf-8", response.getHeader("Content-Type").getValue());
// ... more asserts
```

so explicitly setting charset=utf-8 works already (unless an error servlet works any different than a normal servlet - which I doubt)

The only potential change would be to *opt-in* for keeping the old behavior in which `res.setCharacterEncoding("UTF-8");` enforces adding `;charset=utf-8` to `application/json`, so that older web-apps behave the same without modifications. I believe this has some value, but I understand if you don't... and a workaround with a servlet Filter can be done as you mentioned.

**Comment by joakime** [comment]
> @joakime what is the argument for us rewriting this header?
> We don't correct "text/plane" to "text/plain", so why should we correct other mistakes?

I'm totally on-board with this argument.
But only for the 2 cases where the entire `Content-Type` response header is set ...

Either directly, as a header via ...

``` java
HttpServletResponse.setHeader("Content-Type", "application/json;charset=utf-8");
```

or indirectly via the API ...

``` java
HttpServletResponse.setContentType("application/json;charset=utf-8");
```

But not via the split header API usage ...
``` java
// this usage should still use the inferred / assumed charset definitions from our MimeTypes
HttpServletResponse.setCharacterEncoding("ISO-8859-1");
HttpServletResponse.setContentType("text/plain"); // results in "text/plain;charset=iso-8859-1"

// and this usage ...
HttpServletResponse.setCharacterEncoding("UTF-8");
HttpServletResponse.setContentType("application/json"); // results in only "application/json" due to assumed charset rules
```

**Comment by joakime** [comment]
@gregw I noticed we don't have a means to programatically adjust the predefined assumed / inferred encodings in our MimeTypes object.

For example, someone couldn't come along and remove the assumed charset mapping for `application/json` in our current MimeTypes.Mutable interface, perhaps we should have a `removeAssumed(String contentType)` and `removeInferred(String contentType)` methods too, so that this kind of behavior can just be adjusted at runtime too.

**Comment by joakime** [comment]
@gregw also, what about people that use the split API and also use content-types that don't have a charset? (like images or binary)

``` java
// this should never have a charset show up in the Content-Type field.
HttpServletResponse.setCharacterEncoding("UTF-8");
HttpServletResponse.setContentType("image/png");
```

It could be that a filter blanket calls the `.setCharacterEncoding("UTF-8")` (glaring at jersey and spring), but then when it reaches the endpoint Servlet, the servlet only calls `.setContentType("image/png")`, which never has a charset.

We now have a split API usage with a charset where it doesn't make sense.

**Comment by gregw** [comment]
@joakime maybe we need assumed, inferred and ignored charsets in our MimeTypes?

**Comment by joakime** [comment]
> @joakime maybe we need assumed, inferred and ignored charsets in our MimeTypes?

I can't think of a better approach yet.

**Comment by gregw** [comment]
@joakime I'm inclined to close this as nothing to do?

**Comment by gregw** [comment]
I've done a speculative PR to make assumed charsets override previously set ones.... but I'm not 100% convinced we want to make that change.

**Comment by joakime** [comment]
Closing as PR #12986
